AN(C)OVA Contrasts now uses emmeans/afex

JohnnyDoorn opened the pull request:
This plays a lot nicer with custom contrasts:
- less restrictions on custom contrasts
- will be very easy to add contrasts on interactions (now or in the next release)
- feedback about contrasts is now on the level of contrast container instead of whole anova container
I did have to sacrifice the option to not assume equal variances. We can add this back later, but it will be tricky (i..e, by having the whole anova be a gls model). I would argue it's a bit of a weird option anyway, since you want equal variances in ANOVA.

Commit:
hackhack

Commit:
contrasts use afex

Commit:
Removed option to assume equal variances for contrasts - contrast analysis now uses emmeans package

Commit:
Update helpfiles

Commit:
Remove redundant code

Commit:
better error message RM ANOVA eq var

Commit:
remove browser comment

vandenman commented on JASP-Engine/JASP/R/anovarepeatedmeasures.R:
```diff
@@ -1252,6 +1245,12 @@ AnovaRepeatedMeasures <- function(jaspResults, dataset = NULL, options) {
+        
+      } else if (options$contrastAssumeEqualVariance == FALSE) {
+        
+        contrastContainer[[paste0(contrast$contrast, "Contrast_",  contrast$variable)]]$setError(gettextf("Unequal variances only available for within subjects factors"))
```
```suggestion
        contrastContainer[[paste0(contrast$contrast, "Contrast_",  contrast$variable)]]$setError(gettext("Unequal variances only available for within subjects factors"))
```
`gettext` rather than `gettextf`.

vandenman commented on JASP-Engine/JASP/R/anovarepeatedmeasures.R:
```diff
@@ -1204,26 +1203,19 @@ AnovaRepeatedMeasures <- function(jaspResults, dataset = NULL, options) {
+      if (contrast$contrast == "custom") {
+        if (isTryError(contrastResult)) {
+          if (grepl(contrastResult[1], pattern = "Nonconforming number")) {
+            contrastContainer$setError(gettextf("Nonconforming number of contrast coefficients."))
```
```suggestion
            contrastContainer$setError(gettext("Nonconforming number of contrast coefficients."))
```

vandenman commented on JASP-Engine/JASP/R/anovarepeatedmeasures.R:
```diff
@@ -1204,26 +1203,19 @@ AnovaRepeatedMeasures <- function(jaspResults, dataset = NULL, options) {
+          if (grepl(contrastResult[1], pattern = "Nonconforming number")) {
+            contrastContainer$setError(gettextf("Nonconforming number of contrast coefficients."))
+          } else if (grepl(contrastResult[1], pattern = "number of contrast matrix rows")) {
+            contrastContainer$setError(gettextf("Wrong number of custom contrast matrix rows."))
```
```suggestion
            contrastContainer$setError(gettext("Wrong number of custom contrast matrix rows."))
```

vandenman commented on JASP-Engine/JASP/R/anovarepeatedmeasures.R:
```diff
@@ -1006,7 +1007,7 @@ AnovaRepeatedMeasures <- function(jaspResults, dataset = NULL, options) {
     
     if (makeBootstrapTable) {
-      thisOverTitle <- gettextf("%1$s%% bca%2$s CI", options$confidenceIntervalIntervalPostHoc * 100, "\u002A")
+      thisOverTitle <- gettextf("%1$s%% bca%2$s CI", options$confidenceIntervalIntervalPostHoc * 100, "\u2020")
```
perhaps add:

```r
contrastTable$addFootnote(symbol = "\u2020", message = gettext("Bias corrected accelerated."))
```

so people don't need to look/ enable other tables to know what `bca` stands for.

JohnnyDoorn replied:
That was already in there, but I moved it to that same if statement (there was the same if statement later to added some more stuff based on makeboostraptable).

vandenman commented on JASP-Engine/JASP/R/anovarepeatedmeasures.R:
```diff
@@ -257,13 +257,11 @@ AnovaRepeatedMeasures <- function(jaspResults, dataset = NULL, options) {
-  variables <- unlist(c(options$betweenSubjectFactors, lapply(options$repeatedMeasuresFactors, function(x) x$name)))
-
-  for (variable in variables)
+  # variables <- unlist(c(options$betweenSubjectFactors, lapply(options$repeatedMeasuresFactors, function(x) x$name)))
```
delete?

vandenman commented on JASP-Engine/JASP/R/ancova.R:
```diff
@@ -1247,7 +1128,7 @@ Ancova <- function(jaspResults, dataset = NULL, options) {
     
     if (options$confidenceIntervalsPostHoc) {
-      thisOverTitle <- gettextf("%g%% CI for Mean Difference", options$confidenceIntervalIntervalPostHoc * 100)
+      thisOverTitle <- gettextf("% CI for Mean Difference", options$confidenceIntervalIntervalPostHoc * 100)
```
```suggestion
      thisOverTitle <- gettextf("%s%% CI for Mean Difference", options$confidenceIntervalIntervalPostHoc * 100)
```
otherwise, `options$confidenceIntervalIntervalPostHoc * 100` won't be inserted into the string.

JohnnyDoorn replied:
Ah yea, haven't really worked with those gettext functions before...

vandenman commented on JASP-Engine/JASP/R/ancova.R:
```diff
@@ -719,59 +621,38 @@ Ancova <- function(jaspResults, dataset = NULL, options) {
+            contrastContainer$setError(gettextf("Wrong number of custom contrast matrix rows."))
+          }
+          return()
+        } else if (any(apply(contrastMatrix, 2, function(x) all(x == 0) ))) {
```
maybe
```suggestion
        } else if (any(colSums(contrastMatrix) == 0)) {
```
?

JohnnyDoorn replied:
Yeah, way better!

vandenman commented on JASP-Engine/JASP/R/ancova.R:
```diff
@@ -719,59 +621,38 @@ Ancova <- function(jaspResults, dataset = NULL, options) {
+          if (grepl(contrastResult[1], pattern = "Nonconforming number")) {
+            contrastContainer$setError(gettextf("Nonconforming number of contrast coefficients."))
+          } else if (grepl(contrastResult[1], pattern = "number of contrast matrix rows")) {
+            contrastContainer$setError(gettextf("Wrong number of custom contrast matrix rows."))
```
```suggestion
            contrastContainer$setError(gettext("Wrong number of custom contrast matrix rows."))
```

vandenman commented on JASP-Engine/JASP/R/ancova.R:
```diff
@@ -719,59 +621,38 @@ Ancova <- function(jaspResults, dataset = NULL, options) {
+      if (contrast$contrast == "custom") {
+        if (isTryError(contrastResult)) {
+          if (grepl(contrastResult[1], pattern = "Nonconforming number")) {
+            contrastContainer$setError(gettextf("Nonconforming number of contrast coefficients."))
```
```suggestion
            contrastContainer$setError(gettext("Nonconforming number of contrast coefficients."))
```

vandenman commented on JASP-Engine/JASP/R/ancova.R:
```diff
@@ -648,33 +553,35 @@ Ancova <- function(jaspResults, dataset = NULL, options) {
-      contrastTable$addColumnInfo(name="lwrBound", type = "number", title = gettext("Lower"), overtitle = thisOverTitle, )
-      contrastTable$addColumnInfo(name="uprBound", type = "number", title = gettext("Upper"), overtitle = thisOverTitle)
+      thisOverTitle <- gettextf("%s%% CI for Mean Difference", options$confidenceIntervalIntervalContrast * 100)
+      contrastTable$addColumnInfo(name="lower.CL", type = "number", title = gettext("Lower"), overtitle = thisOverTitle)
```
Renaming is fine, but what does "CL" stand for? It's not a Confidence Interval?

JohnnyDoorn replied:
I guess confidence limit or something? It's how the confint function, applied to class emmGrid, returns the confidence intervals.

Commit:
Don feedback

Commit:
Some fixes already to failed unit test

JohnnyDoorn commented:
Probably some issues still with a different errorchecking - @vandenman did anything change to the BANOVA error checks?

vandenman commented:
Well there were some changes to `.hasErrors` (https://github.com/jasp-stats/jasp-desktop/pull/3855) but aside from that, not that I know.

Commit:
fixes to unit test, update of helpfiles
RM ANOVA can now use multivariate model for follow up tests (recommended by Henrik)
